file_writer_control/CommandChannel.py

```diff
@@ -0,0 +1,217 @@
+            self.map_of_commands[command_id] = CommandStatus(job_id, command_id)
+
+    def stop_thread(self):
+        """
+        Stop the thread that is continuously getting command topic messages in the background. Should only be called if
+        we are about to get rid of the current instance of CommandChannel.
+        """
+        self.to_thread_queue.put("exit")
+        try:
+            self.thread.join()
+        except RuntimeError:
+            pass
+
+    def __del__(self):
+        self.stop_thread()
+
+    def update_workers(self):
+        """
+        Update the list of known workers, jobs and commands. This is a non-blocking call but it might take some time
+        to execute if the queue of updates is long. This member function is called by many of the other member functions
+        in this class.
+        """
+        while not self.status_queue.empty():
+            status_update = self.status_queue.get()
+            if type(status_update) is WorkerStatus:
+                if status_update.service_id not in self.map_of_workers:
+                    self.map_of_workers[status_update.service_id] = status_update
+                self.map_of_workers[status_update.service_id].update_status(
+                    status_update
+                )
+            elif type(status_update) is JobStatus:
+                if status_update.job_id not in self.map_of_jobs:
+                    self.map_of_jobs[status_update.job_id] = status_update
+                self.map_of_jobs[status_update.job_id].update_status(status_update)
+            elif type(status_update) is CommandStatus:
+                if status_update.command_id not in self.map_of_commands:
+                    self.map_of_commands[status_update.command_id] = status_update
+                self.map_of_commands[status_update.command_id].update_status(
+                    status_update
+                )
+            else:
+                pass
+        now = datetime.now()
+        for worker in self.map_of_workers.values():
+            if worker.last_update + STATUS_MESSAGE_TIMEOUT < now:
+                worker.state = WorkerState.UNAVAILABLE
+        for job in self.map_of_jobs.values():
+            if job.last_update + JOB_STATUS_TIMEOUT < now and not (
+                job.state == JobState.DONE or job.state == JobState.ERROR
+            ):
+                job.state = JobState.TIMEOUT
+        for command in self.map_of_commands.values():
+            if command.last_update + COMMAND_STATUS_TIMEOUT < now and not (
+                command.state == CommandState.SUCCESS
+                or command.state == CommandState.ERROR
+            ):
+                command.state = CommandState.TIMEOUT_RESPONSE
+
+    def list_workers(self) -> List[WorkerStatus]:
+        """
+        :return: A list of the (known) workers with state and status information.
+        """
+        self.update_workers()
+        return list(self.map_of_workers.values())
```

Comment:
Why are we calling the `list()` constructor instead of `[]`? The constructor calls are very expensive in Python unless otherwise they are surely needed.

Reply:
Because `list(dict_values) != [dict_values]`. Thus it is needed in this particular case.

Reply:
In what way they are not equal?